Ensure CookieStore uses the creation URL for setting/getting cookies

RupinMittal opened the pull request:
#### 8a2b852c5d38ea4ecae01783031902f7a99d3a8b
<pre>
Ensure CookieStore uses the creation URL for setting/getting cookies
<a href="https://bugs.webkit.org/show_bug.cgi?id=279454">https://bugs.webkit.org/show_bug.cgi?id=279454</a>
<a href="https://rdar.apple.com/135730639">rdar://135730639</a>

Reviewed by Brady Eidson.

Per the Cookie Store API spec, (<a href="https://wicg.github.io/cookie-store/)">https://wicg.github.io/cookie-store/)</a>, when setting or
getting cookies, the CookieStore should use the url passed in or setting&apos;s creation url.
Up till now, we were using setting&apos;s current url. This is a problem because the current
url can change--meaning that if the current url changes, then requests to get cookies
set using the preivous url will fail. The current WPT tests don&apos;t test this scenario.

This patch updates set(), get(), and getAll() to use the url passed in or the creation url.
We also add two layout tests to test the case where a cookie is set, the document&apos;s
current URL changes, and then there is a request to get the cookies. These tests fail
without the changes to set(), get(), and getAll().

There will be a future patch to the WPT repository to add the tests there.

Since the CookieStoreGetOptions passed into get() and getAll() are still sent over IPC
but their `url` field is not used, we set this field to a null string so that we are not
sending data unnecessarily. We still want to send the options because this object may
change in the future, so it&apos;s better practice to have the code sending it already.

Note that this causes the layout test cookieStore_opaque_origin.https.html to fail. This
test attempts to set a cookie from an iframe. Before this change, this succeeded because
CookieStore::get() was incorrectly using the host obtained from the iframe&apos;s security
origin. Since the iframe didn&apos;t have it&apos;s own security origin, it was given the security
origin of it&apos;s parent--so the test was able to obtain the correct host. But this is the
wrong method of obtaining the host. As per the spec, the host should be obtained from the
url. Now that we use the creation URL (which is about:srcdoc for this iframe), the host
is null, and the test fails.

This breakage in iframe (also occurring in cookieStore_get_set_across_frames.https.html)
will be fixed in the next patch that will travel up the parent tree until a valid creation
url is found. I didn&apos;t do this in this patch so that this patch could be small and focus
on switching set(), get(), and getAll() to using the creation url.

* LayoutTests/imported/w3c/web-platform-tests/cookie-store/cookieStore_getAll_set_creation_url.https.any-expected.txt: Added.
* LayoutTests/imported/w3c/web-platform-tests/cookie-store/cookieStore_getAll_set_creation_url.https.any.html: Added.
* LayoutTests/imported/w3c/web-platform-tests/cookie-store/cookieStore_getAll_set_creation_url.https.any.js: Added.
(promise_test.async testCase.async let):
(promise_test.async testCase):
* LayoutTests/imported/w3c/web-platform-tests/cookie-store/cookieStore_get_set_across_frames.https-expected.txt:
* LayoutTests/imported/w3c/web-platform-tests/cookie-store/cookieStore_get_set_creation_url.https.any-expected.txt: Added.
* LayoutTests/imported/w3c/web-platform-tests/cookie-store/cookieStore_get_set_creation_url.https.any.html: Added.
* LayoutTests/imported/w3c/web-platform-tests/cookie-store/cookieStore_get_set_creation_url.https.any.js: Added.
(promise_test.async testCase.async let):
(promise_test.async testCase):
* LayoutTests/imported/w3c/web-platform-tests/cookie-store/cookieStore_opaque_origin.https-expected.txt:
* Source/WebCore/Modules/cookie-store/CookieStore.cpp:
(WebCore::CookieStore::MainThreadBridge::get):
(WebCore::CookieStore::MainThreadBridge::set):
(WebCore::CookieStore::get):
(WebCore::CookieStore::getAll):
(WebCore::CookieStore::set):
* Source/WebCore/dom/Document.h:
(WebCore::Document::creationURL const): Deleted.
* Source/WebCore/dom/EmptyScriptExecutionContext.h:
* Source/WebCore/dom/ScriptExecutionContext.h:
* Source/WebCore/workers/WorkerGlobalScope.h:
* Source/WebCore/worklets/WorkletGlobalScope.h:

Canonical link: <a href="https://commits.webkit.org/283530@main">https://commits.webkit.org/283530@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/33e4a8bccfda182579878434ca757612490fd2ba

| Misc | iOS, visionOS, tvOS & watchOS  | macOS  | Linux |  Windows |
| ----- | ---------------------- | ------- |  ----- |  --------- |
| [✅ 🧪 style](https://ews-build.webkit.org/#/builders/38/builds/66542 "Passed style check") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/48/builds/45917 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/55/builds/19163 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/5/builds/70576 "Built successfully") | [✅ 🛠 wincairo](https://ews-build.webkit.org/#/builders/59/builds/17675 "Built successfully") 
| [✅ 🧪 bindings](https://ews-build.webkit.org/#/builders/9/builds/68660 "Passed tests") | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/49/builds/53716 "Built successfully") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/61/builds/17435 "Built successfully") | [✅ 🧪 wpe-wk2](https://ews-build.webkit.org/#/builders/34/builds/53334 "Passed tests") | [✅ 🧪 wincairo-tests](https://ews-build.webkit.org/#/builders/60/builds/11922 "Passed tests") 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/11/builds/69609 "Passed tests") | [✅ 🧪 ios-wk2](https://ews-build.webkit.org/#/builders/47/builds/42282 "Passed tests") | [✅ 🧪 api-mac](https://ews-build.webkit.org/#/builders/18/builds/57584 "Passed tests") | [✅ 🧪 api-wpe](https://ews-build.webkit.org/#/builders/41/builds/33997 "Passed tests") | 
| | [✅ 🧪 ios-wk2-wpt](https://ews-build.webkit.org/#/builders/42/builds/38953 "Passed tests") | [✅ 🧪 mac-wk1](https://ews-build.webkit.org/#/builders/64/builds/14972 "Passed tests") | [✅ 🛠 wpe-cairo](https://ews-build.webkit.org/#/builders/65/builds/16029 "Built successfully") | 
| | [✅ 🧪 api-ios](https://ews-build.webkit.org/#/builders/13/builds/60863 "Passed tests") | [✅ 🧪 mac-wk2](https://ews-build.webkit.org/#/builders/63/builds/15314 "Passed tests") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/2/builds/72278 "Built successfully") | 
| | [✅ 🛠 vision](https://ews-build.webkit.org/#/builders/87/builds/10499 "Built successfully") | [✅ 🧪 mac-AS-debug-wk2](https://ews-build.webkit.org/#/builders/62/builds/14682 "Passed tests") | [✅ 🧪 gtk-wk2](https://ews-build.webkit.org/#/builders/1/builds/60657 "Passed tests") | 
| | [✅ 🛠 vision-sim](https://ews-build.webkit.org/#/builders/86/builds/10531 "Built successfully") | [✅ 🧪 mac-wk2-stress](https://ews-build.webkit.org/#/builders/8/builds/57653 "Passed tests") | [✅ 🧪 api-gtk](https://ews-build.webkit.org/#/builders/21/builds/60980 "Passed tests") | 
| | [✅ 🧪 vision-wk2](https://ews-build.webkit.org/#/builders/88/builds/8643 "Passed tests") | [✅ 🧪 mac-intel-wk2](https://ews-build.webkit.org/#/builders/119/builds/2261 "Passed tests") | | 
| [✅ 🛠 🧪 unsafe-merge](https://ews-build.webkit.org/#/builders/22/builds/10089 "Built successfully and passed tests") | [  ~~🛠 tv~~](https://ews-build.webkit.org/#/builders/44/builds/41724 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | | | 
| | [✅ 🛠 tv-sim](https://ews-build.webkit.org/#/builders/46/builds/42801 "Built successfully") | | | 
| | [✅ 🛠 watch](https://ews-build.webkit.org/#/builders/43/builds/43984 "Built successfully") | | | 
| | [✅ 🛠 watch-sim](https://ews-build.webkit.org/#/builders/45/builds/42544 "Built successfully") | | | 
<!--EWS-Status-Bubble-End-->

beidson commented on Source/WebCore/Modules/cookie-store/CookieStore.cpp:
```diff
@@ -383,8 +384,10 @@ void CookieStore::set(CookieInit&& options, Ref<DeferredPromise>&& promise)
     static constexpr auto maximumAttributeValueSize = 1024;
+
+    auto url = context->creationURL();
+    auto host = url.host().toString();
```
Should just keep host as a StringView - We make a StringView out of it later, even!

RupinMittal replied:
Good point--fixed!

beidson reviewed (approved):
R+ with a nit

RupinMittal committed:
Ensure CookieStore uses the creation URL for setting/getting cookies
https://bugs.webkit.org/show_bug.cgi?id=279454
rdar://135730639

Reviewed by Brady Eidson.

Per the Cookie Store API spec, (https://wicg.github.io/cookie-store/), when setting or
getting cookies, the CookieStore should use the url passed in or setting's creation url.
Up till now, we were using setting's current url. This is a problem because the current
url can change--meaning that if the current url changes, then requests to get cookies
set using the preivous url will fail. The current WPT tests don't test this scenario.

This patch updates set(), get(), and getAll() to use the url passed in or the creation url.
We also add two layout tests to test the case where a cookie is set, the document's
current URL changes, and then there is a request to get the cookies. These tests fail
without the changes to set(), get(), and getAll().

There will be a future patch to the WPT repository to add the tests there.

Since the CookieStoreGetOptions passed into get() and getAll() are still sent over IPC
but their `url` field is not used, we set this field to a null string so that we are not
sending data unnecessarily. We still want to send the options because this object may
change in the future, so it's better practice to have the code sending it already.

Note that this causes the layout test cookieStore_opaque_origin.https.html to fail. This
test attempts to set a cookie from an iframe. Before this change, this succeeded because
CookieStore::get() was incorrectly using the host obtained from the iframe's security
origin. Since the iframe didn't have it's own security origin, it was given the security
origin of it's parent--so the test was able to obtain the correct host. But this is the
wrong method of obtaining the host. As per the spec, the host should be obtained from the
url. Now that we use the creation URL (which is about:srcdoc for this iframe), the host
is null, and the test fails.

This breakage in iframe (also occurring in cookieStore_get_set_across_frames.https.html)
will be fixed in the next patch that will travel up the parent tree until a valid creation
url is found. I didn't do this in this patch so that this patch could be small and focus
on switching set(), get(), and getAll() to using the creation url.

* LayoutTests/imported/w3c/web-platform-tests/cookie-store/cookieStore_getAll_set_creation_url.https.any-expected.txt: Added.
* LayoutTests/imported/w3c/web-platform-tests/cookie-store/cookieStore_getAll_set_creation_url.https.any.html: Added.
* LayoutTests/imported/w3c/web-platform-tests/cookie-store/cookieStore_getAll_set_creation_url.https.any.js: Added.
(promise_test.async testCase.async let):
(promise_test.async testCase):
* LayoutTests/imported/w3c/web-platform-tests/cookie-store/cookieStore_get_set_across_frames.https-expected.txt:
* LayoutTests/imported/w3c/web-platform-tests/cookie-store/cookieStore_get_set_creation_url.https.any-expected.txt: Added.
* LayoutTests/imported/w3c/web-platform-tests/cookie-store/cookieStore_get_set_creation_url.https.any.html: Added.
* LayoutTests/imported/w3c/web-platform-tests/cookie-store/cookieStore_get_set_creation_url.https.any.js: Added.
(promise_test.async testCase.async let):
(promise_test.async testCase):
* LayoutTests/imported/w3c/web-platform-tests/cookie-store/cookieStore_opaque_origin.https-expected.txt:
* Source/WebCore/Modules/cookie-store/CookieStore.cpp:
(WebCore::CookieStore::MainThreadBridge::get):
(WebCore::CookieStore::MainThreadBridge::set):
(WebCore::CookieStore::get):
(WebCore::CookieStore::getAll):
(WebCore::CookieStore::set):
* Source/WebCore/dom/Document.h:
(WebCore::Document::creationURL const): Deleted.
* Source/WebCore/dom/EmptyScriptExecutionContext.h:
* Source/WebCore/dom/ScriptExecutionContext.h:
* Source/WebCore/workers/WorkerGlobalScope.h:
* Source/WebCore/worklets/WorkletGlobalScope.h:

Canonical link: https://commits.webkit.org/283530@main
